### PR TITLE
Add a function to compute various type flags of a matrix.

### PR DESCRIPTION
**Pull request by glennw**
This provides a more unified way to calculate various properties
of a matrix, rather than has_perspective_component etc.

The preserves_2d_alignment method returns true for transforms
with a 90 degree rotation. This function allows us to differentiate
between that and a pure translation matrix, which will be useful
for some of the upcoming clip / segment code.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/webrender/2218)

### CODE REVIEW

File: webrender/src/util.rs [hunk 1]
```diff
@@ -92,6 +103,40 @@ impl<Src, Dst> MatrixHelpers<Src, Dst> for TypedTransform3D<f32, Src, Dst> {
...
+        let mut mask = MatrixFlags::empty();
+
+        if self.m14 != 0.0 ||
+           self.m24 != 0.0 ||
+           self.m34 != 0.0 ||
+           self.m44 != 1.0 {
+            return MatrixFlags::TRANSLATION |
+                   MatrixFlags::SCALE |
+                   MatrixFlags::AFFINE |
+                   MatrixFlags::PERSPECTIVE;
```
Comment (kvark) [inline webrender/src/util.rs]: can we really say the matrix has translation/scale if we only check for the perspective components here?

File: webrender/src/util.rs [hunk 2]
```diff
@@ -92,6 +103,40 @@ impl<Src, Dst> MatrixHelpers<Src, Dst> for TypedTransform3D<f32, Src, Dst> {
...
+        if self.m41 != 0.0 ||
+           self.m42 != 0.0 ||
+           self.m43 != 0.0 {
+            mask |= MatrixFlags::TRANSLATION;
+        }
+
+        if self.m11 != 1.0 ||
+           self.m22 != 1.0 ||
+           self.m33 != 1.0 {
+            mask |= MatrixFlags::SCALE;
```
Comment (kvark) [inline webrender/src/util.rs]: could also be pure rotation without scale

**Comment by glennw** [comment]
The flags returned are conservative - this is the comment in the skia source:

```
Returns a bit field describing the transformations the matrix may
        perform. The bit field is computed conservatively, so it may include
        false positives. For example, when kPerspective_Mask is set, all
        other bits are set.
```

**Comment by kvark** [comment]
So what you are trying to encode is 
```rust
enum MatrixKind {
  Simple { scale: bool, translation: bool },
  Affine,
  Perspective,
}
```
Would it be more convenient for the user code to work with flags rather than enums?

**Comment by glennw** [comment]
It's not clear to me at this point if it would be easier for user code to use the flags or the enum - I do like the enum approach in general. We could probably either merge this now, or I can close this and incorporate it as part of a larger PR which will make use of it.

**Comment by kvark** [comment]
@glennw sounds good to me. When you have a use case, you'd have a second look at the enum vs flag and make this call.

**Comment by glennw** [comment]
I realized that what I *actually* want in this use case is going to be covered by the ```TransformOrOffset``` in the PR that @mrobinson is working on. So I'll close that and base my code on that PR.

**Pull request closed without merging**